File: SceneViewSwift/Sources/SceneViewSwift/FrameRate/FrameRatePolicy.swift
#if os(iOS) || os(macOS) || os(visionOS)

/// How often a ``SceneView`` does its per-frame work, and what display cadence it asks for.
///
/// Mirrors SceneView Android's `FrameRatePolicy`
/// (`sceneview/src/main/java/io/github/sceneview/FrameRatePolicy.kt`), case for case and
/// name for name. Pass it with ``SceneView/frameRatePolicy(_:)``:
///
/// ```swift
/// SceneView { /* … */ }                                   // .onDemand() — the default
/// SceneView { /* … */ }.frameRatePolicy(.continuous())    // tick every vsync, never park
/// SceneView { /* … */ }.frameRatePolicy(.onDemand(maxFps: 30))
/// ```
///
/// ## Read this before you believe the name
///
/// On Android the render loop belongs to SceneView: it calls Filament's
/// `beginFrame` / `render` / `endFrame` itself, so `OnDemand` can decide not to draw and
/// **no GPU frame happens**.
///
/// On iOS it does not. `SceneView` is a SwiftUI `RealityView`, and RealityKit owns its own
/// render loop — there is no public gate, no `isPaused`, no `setNeedsDisplay`, and
/// `ARView.RenderOptions` carries no cadence option. **`.onDemand` therefore cannot stop
/// RealityKit from presenting a frame**, and this type does not pretend otherwise.
///
/// What it *does* control on iOS is real, and it is the whole of what the platform offers:
///
/// 1. **SceneViewSwift's own per-frame driver.** The camera-motion loop — the one writer for
///    motion no finger is driving (the coast a released drag leaves, the ``SceneView/autoRotate(speed:)``
///    turntable) — runs on a `CADisplayLink` under this policy. Under ``onDemand(maxFps:)``
///    it stops once nothing is moving and schedules no wake-up at all; under
///    ``continuous(maxFps:)`` it keeps ticking for the view's lifetime.
/// 2. **The display cadence request.** A running driver carries a
///    `CADisplayLink.preferredFrameRateRange`; a parked one is invalidated, so SceneViewSwift
///    asks the system for nothing and a variable-refresh-rate panel is free to idle down.
///    This is the iOS counterpart of Android's `Surface.setFrameRate` vote, and it is a
///    request in exactly the same way — the system may ignore it.
/// 3. **The cap.** ``maxFps`` bounds both the driver's tick rate and the cadence request.
///
/// ## iOS ↔ Android divergence
///
/// | | Android (Filament) | iOS / visionOS (RealityKit) |
/// |---|---|---|
/// | Skip the GPU submit when nothing changed | ✅ the loop owns `beginFrame`/`render`/`endFrame` | ❌ **not possible** — RealityKit presents every vsync |
/// | Park the library's own per-frame work | ✅ | ✅ the camera / coast / turntable driver stops |
/// | Cadence request | ✅ `Surface.setFrameRate` (API 30+) | ✅ `CADisplayLink.preferredFrameRateRange` (iOS 15+); ❌ nothing on macOS |
/// | `maxFps` cap | ✅ presents on whole vsync multiples | ⚠️ caps *our* driver and *our* request — **not** RealityKit's own presentation |
/// | Above 60 Hz on iPhone | (n/a) | ⚠️ needs `CADisableMinimumFrameDurationOnPhone` in the **host app's** `Info.plist`; a framework cannot set it for its host |
/// | `maxFps` ≤ 0 | throws (`require(fps > 0)`) | clamped to `1` — a Swift `enum` case has no failable construction point |
/// | `requestRender()` after mutating a raw object | **required** — the frame would not be drawn | not required for the pixel to appear (RealityKit draws it anyway); use ``SceneRenderInvalidator`` to re-arm the driver and the cadence request |
///
/// The honest one-line summary: on Android `.onDemand` saves the **GPU**; on iOS it saves
/// SceneViewSwift's **CPU** and lets the panel idle, and the GPU is Apple's business.
///
/// ## Choosing
///
/// - ``onDemand(maxFps:)`` — the default, and right for nearly everything. A product viewer
///   nobody is touching stops advancing a camera that is not moving and stops asking for
///   cadence it is not using.
/// - ``continuous(maxFps:)`` — for a scene driven from outside SceneViewSwift every frame
///   (an app-owned animation writing entity transforms on its own timer) where you would
///   rather hold the cadence than call ``SceneRenderInvalidator/requestRender()``.
public enum FrameRatePolicy: Sendable, Hashable {

    /// Do per-frame work only while something is actually happening, then park.
    ///
    /// "Something happening" is a gesture in flight, a drag still coasting, an active
    /// turntable, a framing pass that has not latched, or an explicit
    /// ``SceneRenderInvalidator/requestRender()``. After the last of those the driver still
    /// owes ``FrameRateGate/settleFrames`` ticks before it parks — see that property for why
    /// the iOS reason differs from Android's.
    ///
    /// Parked means parked: the `CADisplayLink` is invalidated, so there is no periodic
    /// wake-up and no cadence request. It does **not** mean the screen stops being redrawn —
    /// RealityKit keeps presenting.
    ///
    /// - Parameter maxFps: ceiling for the driver's tick rate and the cadence request.
    ///   `nil` (the default) asks for the platform maximum. Values below `1` are clamped to `1`.
    case onDemand(maxFps: Int? = nil)

    /// Do per-frame work every vsync for the view's lifetime, and hold the cadence request.
    ///
    /// The driver never parks, even on a scene where nothing moves. This is the closest
    /// analogue of Android's `FrameRatePolicy.Continuous`, and like it, it is the option you
    /// pick when something outside the library changes the scene every frame.
    ///
    /// - Parameter maxFps: ceiling for the driver's tick rate and the cadence request.
    ///   `nil` (the default) asks for the platform maximum. Values below `1` are clamped to `1`.
    case continuous(maxFps: Int? = nil)

    /// The ceiling this policy was constructed with, exactly as written — `nil` for "no cap".
    ///
    /// Mirrors the Kotlin `val maxFps: Int?` on the sealed interface. Use ``resolvedMaxFps``
    /// for the value the driver actually applies, which clamps out the nonsense.
    public var maxFps: Int? {
        switch self {
        case .onDemand(let maxFps), .continuous(let maxFps):
            return maxFps
        }
    }

    /// ``maxFps`` clamped to the range the platform can express: `nil` stays `nil`, anything
    /// below `1` becomes `1`.
    ///
    /// Android rejects `maxFps <= 0` at construction with `require(fps > 0)`. A Swift `enum`
    /// case has no body to throw from, and trapping inside a display-link callback would turn
    /// a caller's typo into a crash in a shipped app, so the Swift port clamps instead and
    /// says so here and in the divergence table on ``FrameRatePolicy``.
    public var resolvedMaxFps: Int? {
        maxFps.map { Swift.max(1, $0) }
    }

    /// Whether the driver keeps ticking with nothing happening.
    ///
    /// `true` only for ``continuous(maxFps:)``. This is the single behavioural bit that
    /// separates the two cases; everything else they share.
    public var ticksWhenIdle: Bool {
        switch self {
        case .onDemand:
            return false
        case .continuous:
            return true
        }
    }
}

#endif

File: sceneview/src/main/java/io/github/sceneview/FrameRatePolicy.kt
package io.github.sceneview

/**
 * How often a [SceneView] submits a frame to the GPU, and what refresh rate it asks the display for
 * while it is doing so.
 *
 * Two independent questions, two independent parts of the type:
 *
 * - **When may a frame be drawn at all?** [OnDemand] draws only when the picture changed;
 *   [Continuous] draws on every vsync.
 * - **How fast, at most?** [maxFps] is an optional ceiling on either of them. `null` — the default
 *   — means the display's own cadence.
 *
 * They compose, which is the point: `OnDemand(maxFps = 30)` is "draw only when something changed,
 * and even then never faster than 30 fps", the most commonly asked-for behaviour and one that no
 * arrangement of three flat cases could express.
 *
 * ```kotlin
 * SceneView { … }                                                  // OnDemand(), the default
 * SceneView(frameRatePolicy = FrameRatePolicy.OnDemand(maxFps = 30)) { … }
 * SceneView(frameRatePolicy = FrameRatePolicy.Continuous()) { … }   // a frame every vsync
 * SceneView(frameRatePolicy = FrameRatePolicy.Continuous(maxFps = 30)) { … }
 * ```
 *
 * The default is [OnDemand]: the library tracks what makes the picture change and draws only then.
 * This replaces the `isRendering` flag, which asked every caller to compute that answer by hand.
 *
 * Whatever the policy, the Compose frame loop itself keeps its normal cadence for everything that is
 * not a GPU submit — async loads still progress, node ticks still run. A policy only decides whether
 * `Renderer.beginFrame` / `render` / `endFrame` is reached on a given tick.
 *
 * **`ARSceneView` takes no policy.** A live camera feed is never idle, so there is nothing to park:
 * its loop runs every vsync, updates the ARCore session every vsync, and skips only the GPU submit,
 * on a tick where ARCore hands back a duplicate `Frame.timestamp` *and* nothing in the virtual scene
 * changed. Tracking, anchors and plane detection are unaffected.
 *
 * @see SceneView
 * @see rememberRenderInvalidator
 */
sealed interface FrameRatePolicy {

    /**
     * The highest frame rate this scene may present at, or `null` (the default) for the display's
     * own cadence.
     *
     * A cap can only ever be met by presenting on a **whole number of vsyncs**: nothing else exists
     * to present on. So the requested period is rounded to whole vsyncs of the display's *real*
     * period, and rounded **up**, because the promise is "never faster than [maxFps]".
     * `maxFps = 90` on a 120 Hz panel therefore runs at 60, not at 120: 90 is not reachable there,
     * and of the two reachable neighbours only 60 honours the promise.
     *
     * The cadence vote sent to the display follows it: a capped scene asks the panel for [maxFps]
     * rather than for its maximum, so a variable-refresh-rate panel can settle on a lower mode
     * instead of running at its ceiling to serve frames that will not come.
     *
     * Must be strictly positive when present — a cap of zero or less is a scene that never draws,
     * which is not what any caller means, so it is rejected at construction rather than silently
     * freezing the view.
     */
    val maxFps: Int?

    /**
     * Render only when something actually changed, then stay parked. **The default.**
     *
     * An idle scene costs no GPU frame and no periodic CPU wake-up — the loop suspends on the
     * snapshot rather than polling — and it wakes on its own for every source of change the library
     * knows about: a touch, a camera manipulator still coasting, a playing animation, a smooth
     * transform, a node added / moved / removed, a visibility, geometry or material change, an async
     * model or environment load, a video frame arriving on a `VideoNode`, a surface resize, a
     * lifecycle resume, an active `surfaceMirrorer`. After the last change it keeps drawing for a
     * short settle window, because Filament finalises texture uploads, IBL and shadow work across
     * several frames.
     *
     * What it never treats as a change is a **recomposition**: an invalidation comes from the thing
     * that changed, not from the fact that the composable ran again. Write your own screen the same
     * way — a counter that writes Compose state from `onFrame` recomposes once per presented frame
     * for no reason, and is measuring itself rather than the scene.
     *
     * Mutating Filament objects directly — a material parameter, a light property written through
     * `LightManager`, a `Skybox` or `IndirectLight` put straight on the Filament `Scene`, morph
     * weights or bone transforms through `RenderableManager`, an external `Stream`, runtime `View`
     * options — happens below the library's bookkeeping and is the one case it cannot see. Call
     * [io.github.sceneview.node.Node.requestRender] or [RenderInvalidator.requestRender] after such
     * an edit; before a `PixelCopy` / screenshot of the surface, request the frame and then wait for
     * your next `onFrame`, because the request is fire-and-forget.
     *
     * @param maxFps optional ceiling — see [FrameRatePolicy.maxFps]. With a cap, a change still
     * wakes the scene immediately; the cap only spaces out the frames drawn while it is awake.
     */
    data class OnDemand(override val maxFps: Int? = null) : FrameRatePolicy {
        init {
            requirePositiveMaxFps(maxFps)
        }
    }

    /**
     * Render on every vsync for as long as the view is composed, and hold a cadence vote the whole
     * time. This is what every `SceneView` did before [OnDemand] became the default.
     *
     * Reach for it when the scene is driven by something outside the library's knowledge and you do
     * not want to invalidate by hand — an external simulation writing into Filament each frame, a
     * custom `Renderer` hook, a texture updated off-thread.
     *
     * `Continuous(maxFps = 30)` is the deliberate-cadence case: a 30 fps product turntable on a
     * 120 Hz panel, drawn whether or not the library thinks anything changed.
     *
     * @param maxFps optional ceiling — see [FrameRatePolicy.maxFps].
     */
    data class Continuous(override val maxFps: Int? = null) : FrameRatePolicy {
        init {
            requirePositiveMaxFps(maxFps)
        }
    }
}

private fun requirePositiveMaxFps(maxFps: Int?) {
    require(maxFps == null || maxFps > 0) {
        "FrameRatePolicy.maxFps must be null (the display's cadence) or strictly positive, was $maxFps"
    }
}
